### DISCUSSION TITLE
Enabling MIDI: USB and/or BT

### DISCUSSION THREAD

**Discussion by thuntjazz** [cardonabits/haxo-hw#47] [Q&A] [answered]
Hi, everyone:

I've been having fun with my Haxophone since it arrived a few days ago!  Big thanks to Javier & contributors for your work.

I'm approaching this project as a saxophonist/wind synth player firstly; I'm a major n00b when it comes to RPi.  So I need some help getting the coding right to enable MIDI via USB and/or Bluetooth.  I'm running a Zero W v1.1.

First off: RPi login - I've connected to a monitor & keyboard, but I'm stuck at the login/password stage.  I've tried both pi and haxophone for the login, and both raspberry and leaving blank for the password, but no joy so far.

Then I guess I need a tutorial to get my MacBook to recognize the Haxophone. It's not recognized either plugged in to USB or with BT.

Finally, I need to get the MIDI out working.

My goal is to set up the Haxophone not only as a standalone, but also as a controller for softsynths, and possibly even hardware synths with the right cabling.  Even better would be Bluetooth MIDI--I'm a big fan of the CME WIDI products, and I use them successfully on my Akai EWI and my Yamaha WX5.

Thanks!

**Comment by jcard0na** [comment]
Hi @thuntjazz!

Glad to hear from a real saxophonist!

Regarding MIDI, the good news is that you do not need to do much to get it to work on a Mac.  Just plug your USB cable on the **inner** usb port of the Rasbperry Pi Zero (Labelled "USB", not the one labelled "PWR IN").  Once you do that, it should be recognized by your Mac as a midi instrument.  Check out [this update](https://www.crowdsupply.com/cardona-bits/haxophone/updates/announcing-were-104-funded-and-usb-midi-support#even-more-thrilling-news-introducing-usb-midi-support) to see this demonstrated.

By the way, if you still want to log in, you were very close... the username is `pi` and the password is `haxophone`

Looking forward to see what you do with the Haxophone and other synths!

Best,

**Reply by thuntjazz** [reply]
Hi, Javier:

Thanks for your super quick reply!  It's 0130 here and I can't put the thing down  :-)

I did get logged in and played around with settings enough where I was able to pair Bluetooth with both my Mac and my Android phone.  Once I shut down and plugged in without peripherals and only connected to power, however...no connection.  I'm wondering if my config needs to be set for an autologin, then maybe BT will also connect automatically?  That's tomorrow's attempt.

Still no luck getting MIDI signal out from the pi, though.  I've tried plugging in just as your video showed, using several cables into different ports, powered hubs, etc...nothing.  I also tried using a wireless MIDI host that is proven to work with my EWI USB...nothing there either.  

I really think the pi just is not sending MIDI signal.  Do any of the tests you've run in development check for MIDI signal out?  I may have to get my hands on another pi to see if it's a hardware issue.

It does play great when it's just powered up by itself with the headphone out!  But it would be super cool to be able to use it as a controller in a DAW...besides just being neat, that was my primary reason for buying a Haxophone in the first place.

Time for bed...we'll try again tomorrow!

Tom

**Reply by jcard0na** [reply]
:man_facepalming: The current firmware version that is available through the [bit.ly link](https://bit.ly/haxo-rpiz) contains an older version of haxo-rs.  I will fix this later today, or if you feel adventurous, you can build it yourself:  https://github.com/cardonabits/haxo-hw/blob/main/docs/development.md#compilation-method-1-native-compilation-on-the-raspi

My apologies, and I hope this blunder was not the only reason for you to stay up until 1:30! :disappointed:

**Reply by jcard0na** [reply, marked as answer]
OK, [here you are](http://94.143.139.181/haxophone-rpiz.img.zip).  This image has USB enabled.  It now enumerates properly as a Haxophone, so it will make it easier to recognize if you have other MIDI instruments plugged in.  I tested on Mac and Linux, but briefly.  Once I've tested it more I'll announce it more widely.  For now, I'm just sharing it here, so you don't say up too late a second night in a row trying to figure this out :wink:

**Reply by thuntjazz** [reply]
It took 2 attempts to load the image and get the Haxophone to respond, but it works!  There is definitely signal coming out the USB now - the Mac recognizes it.  In the next day or two, I'll try my WIDI Uhost with a hardware synth - I'm curious to see how well that works.

In the meantime, I'm going to bed much earlier tonight  :-)

Thanks a bunch - great work!

**Comment by JackWitherell** [comment]
joining in on this discussion, just got my haxophone, really like it.

I'd really like to know if there's a way to enable midi via a generic usb midi adapter on the pi 4. It'd be really great to be able to interface this with DIN midi instead of having to source a midi host (like a computer) for using with non usb gear. Has this been explored? is the generic USB midi adapter discovery too vague from a software perspective?

**Reply by jcard0na** [reply]
Hi Jack,

I have not explored it.  My experience with generic USB adapters is limited to using them on a MIDI host.  I suppose one can also use them to output MIDI.  In that case, I see no problem in routing MIDI events from the haxophone to MIDI out.  I'll be happy to help out if you decide to explore this.

Cheers,

**Reply by thuntjazz** [reply]
Hi, Jack:

There are plenty of USB to 5 pin DIN MIDI adapters out there.  The question is how to get the pi to talk with the interface.  

In my case also, trying to work with a Zero, does a micro USB MIDI interface exist, or can an adapter be used and still transmit the MIDI out accurately?  I've had good luck with CME's wireless, super-low latency MIDI interfaces.  A WIDI Uhost on the pi end and a Widi Master on the hardware end would most likely do the trick.  That's my current setup between my EWI USB and two different hardware synths.  Once I figure out why my Zero isn't transmitting MIDI data, it should work great on the Haxophone as well.

Good luck and keep us posted!

**Reply by JackWitherell** [reply]
> does a micro USB MIDI interface exist

Probably not, definitely not my intention. The second micro USB port can be used with a usb OTG cable to provide a full sized usb port to the pi zero. I was thinking about disabling the current midi method and having the software directly interact with a usb midi IO device. (I didn't realize you could do usb guest from a rpi zero until I saw this project. from the get go, it seemed entirely odd to me how and why midi support on the haxophone was implemented like this in the first place.) I've seen other software that can discover and attach to a generic midi device, which is why I inquired if discovery was a problem or if it was ever explored at all.

Also, I'm much more invested in traditional midi as I'd really like to play my homebuilt monophonic synths like a saxophone. I honestly haven't even tried to test midi out on the haxophone yet since I initially set it up with a pi 4. I've since moved back to pi zero, but I'll likely move back before I experiment further with this, after I at *least* give usb midi mode a shot.

I've heard from the other threads in this haxo forum that latency is an issue, but I don't necessarily see why this should be the case, especially with a relatively low-level connection like midi over usb that the computer component of the rpi can directly interact with and issue midi messages. Specifically breath cc and pitch of course, but for my purposes volume and pitch are my real priorities.
sending MPE messages would be cool but is obviously excessive and not inherently useful (saxaphones are monophonic!)

On the other hand, I do regret not having enough time to dig deeper, my job owns me.

I'm not used to rust but will likely delve into whether or not I can go about issuing midi output to a generic midi interface eventually.

Love the haxophone by the way. Very cool hardware, I'm glad the diy nature could give it space to breathe and hopefully improve.

**Reply by jcard0na** [reply]
> (I didn't realize you could do usb guest from a rpi zero until I saw this project. from the get go, it seemed entirely odd to me how and why midi support on the haxophone was implemented like this in the first place.)

Technically it is called USB On The Go (OTG) and it can act as USB host or device (what you called "guest") depending on what's plugged on the other end.  As you noticed, only the Zeros support that.  The 3 and 4 are just USB host.

>(...)
>Love the haxophone by the way. Very cool hardware, I'm glad the diy nature could give it space to breathe and hopefully improve.

Thanks for the feedback.  There definitely room for improvement... lots of it! 😜

**Comment by Utopiah** [comment]
Hi, is http://94.143.139.181/haxophone-rpiz.img.zip identical to the one in the PDF doc (I believe https://bit.ly/haxo-rpiz )? Because I just tried the bit.ly one and while doing `aplaymidi -l` I don't see the Haxophone (which does work, as I hear the starting chime) while my Akai is seen (while plugged).

**Reply by Utopiah** [reply]
According to the SHA256 they are so I'm not sure why I'm not getting MIDI out.

**Reply by Utopiah** [reply]
Burned the linked image here but no MIDI on my RPi Zero 1 W unfortunately, at least not according to aplaymidi.

**Reply by jcard0na** [reply]
Sorry, I'm confused.  Is there an image that works and another that does not?  They should be identical...
Anyway, downloading both images now to check.  It will take some time.

**Reply by jcard0na** [reply]
Hi,

I just checked.  The images are identical
```
sha256sum haxophone-rpiz.img.zip haxophone.img.zip 
377cc1aa4afb0da40eaf4b7b3704f12987064ab0157d2fbf84ef8fda1ea75d23  haxophone-rpiz.img.zip
377cc1aa4afb0da40eaf4b7b3704f12987064ab0157d2fbf84ef8fda1ea75d23  haxophone.img.zip
```

Could it be that you used the power USB port instead of the one labelled `USB`?  USB Midi only works out of that one.

**Reply by Utopiah** [reply]
I use the inside one. What can I do to try to identify what is going wrong?

**Reply by Utopiah** [reply]
I see couple of RPi Zero pins that might be not properly soldered so I'll try that again tomorrow, just in case.

PS: to clarify, the images are indeed the same but as my setup didn't work I was wondering if by any chance they were not and I might have used an outdated image.

**Comment by jcard0na** [comment]
> I use the inside one. What can I do to try to identify what is going wrong?

Pick one of the three methods to [log in](https://github.com/cardonabits/haxo-hw/blob/main/docs/login.md) and look at the logs (`journalctl -u haxo`) to confirm that haxo is running successfully with midi enabled.

On your host run (`sudo dmesg`) and (`lsusb`) to see if it enumerates.

Is regular audio working?

**Reply by Utopiah** [reply]
I know that haxo is running because I can play using headphones but how can I know specifically about MIDI being enabled here and working?

Edit: is it output exclusive, namely either minijack plugged in OR MIDI?

**Reply by jcard0na** [reply]
No, it is not output exclusive: both work concurrently.
If you are using the released images, MIDI is enabled.  You can confirm looking at the logs (there are MIDI specific messages).
Is your host Linux?  Does `dmesg` or `lsusb` not show a USB midi device?

**Reply by Utopiah** [reply]
Here is how it is plugged :

![image](https://github.com/cardonabits/haxo-hw/assets/1076655/157ab35c-e007-43d1-975e-5dedc1510491)

and `lsusb` on the host does not show a new USB MIDI device, in fact it shows no new device at all it seems.

I'm getting 

```
pi@haxophone:~ $ journalctl -u haxo | grep -i midi
Dec 21 18:26:43 haxophone haxo001[508]: [2023-12-21T18:26:43Z INFO  haxo001::midi]
Dec 21 18:26:43 haxophone haxo001[508]: [2023-12-21T18:26:43Z INFO  haxo001::midi] 0: Midi Through:Midi Through Port-0 14:0
Dec 21 18:26:43 haxophone haxo001[508]: [2023-12-21T18:26:43Z INFO  haxo001::midi] 1: f_midi:f_midi 24:0
Dec 21 18:26:43 haxophone haxo001[508]: [2023-12-21T18:26:43Z INFO  haxo001::midi] Picked the last one
Dec 21 20:29:33 haxophone haxo001[504]: [2023-12-21T20:29:33Z INFO  haxo001::midi]
Dec 21 20:29:33 haxophone haxo001[504]: [2023-12-21T20:29:33Z INFO  haxo001::midi] 0: Midi Through:Midi Through Port-0 14:0
Dec 21 20:29:33 haxophone haxo001[504]: [2023-12-21T20:29:33Z INFO  haxo001::midi] 1: f_midi:f_midi 24:0
Dec 21 20:29:33 haxophone haxo001[504]: [2023-12-21T20:29:33Z INFO  haxo001::midi] Picked the last one
Dec 21 20:35:49 haxophone haxo001[633]: [2023-12-21T20:35:49Z INFO  haxo001::midi]
Dec 21 20:35:49 haxophone haxo001[633]: [2023-12-21T20:35:49Z INFO  haxo001::midi] 0: Midi Through:Midi Through Port-0 14:0
Dec 21 20:35:49 haxophone haxo001[633]: [2023-12-21T20:35:49Z INFO  haxo001::midi] 1: f_midi:f_midi 24:0
Dec 21 20:35:49 haxophone haxo001[633]: [2023-12-21T20:35:49Z INFO  haxo001::midi] Picked the last one
```

for the test I did yesterday evening I assume.

**Reply by Utopiah** [reply]
Ugh... f*cking cables! Just tried another random USB cable, just in case, and now see :

`Bus 001 Device 014: ID 1209:2bad Generic Haxophone`

with of course

```

fabien@fabien-CORSAIR-ONE-i160:~$ aplaymidi -l
 Port    Client name                      Port name
 14:0    Midi Through                     Midi Through Port-0
 28:0    Haxophone                        Haxophone MIDI 1

``` 

Well I hope my troubles will help others! Thanks, again, for your precious help!

**Reply by Utopiah** [reply]
Just tried with LMMS and it works :)

**Reply by Utopiah** [reply]
Motivated by https://github.com/cardonabits/haxo-hw/discussions/68#discussioncomment-8076717

**Reply by jcard0na** [reply]
Thanks for letting me know!
Cheers,
